Indirect checking for return_type through an extra function

Keno opened the pull request:
Compiler.return_type is a bit of an odd beast because we basically treat
it like a built-in, but it's defined in the compiler. When working on
inference, it can be useful to load a second copy of inference with
the new changes (e.g. to test changes before the changes are capable
of running correctly on all input). This works quite well, but was
causing problems with return type, because Base's notion of
return_type and the second inference copy's notion of return_type
were different. This patch adds a simple `is_return_type(f)` predicate
that we can overload in our second inference copy in order to make it
recognize both `Core.Compiler.return_type` and `return_type` (from
the second copy's perspective as the return_type builtin).

@jrevels this fixes that weird non-inference of matvec we saw yesterday

vtjnash commented:
> it can be useful to load a second copy of inference with
the new changes

It's definitely wrong to treat a second copy of `Core.Compiler.return_type` as being related in any way to the primary copy. I suppose that's just a question of whether we care that this is UB though.

But, also, this won't (shouldn't?) do anything, since you can't alter the type-infer method table at runtime.

Keno commented:
> It's definitely wrong to treat a second copy of Core.Compiler.return_type as being related in any way to the primary copy. I suppose that's just a question of whether we care that this is UB though.

I don't care if it's wrong as long as it works and is useful ;). Also, I would question your "related in any way" assertion. After all, they implement the same semantics.

> But, also, this won't (shouldn't?) do anything, since you can't alter the type-infer method table at runtime.

I'm not altering the type-infer method table, I'm loading a second copy of inference where `is_return_type` is defined differently.

Keno committed:
Indirect checking for return_type through an extra function

Compiler.return_type is a bit of an odd beast because we basically treat
it like a built-in, but it's defined in the compiler. When working on
inference, it can be useful to load a second copy of inference with
the new changes (e.g. to test changes before the changes are capable
of running correctly on all input). This works quite well, but was
causing problems with return type, because Base's notion of
return_type and the second inference copy's notion of return_type
were different. This patch adds a simple is_return_type(f) predicate
that we can overload in our second inference copy in order to make it
recognize both Core.Compiler.return_type and return_type (from
the second copy's perspective as the return_type builtin).
